File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/TransactionManager.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using Diagnostics;
    using System;
    using System.Fabric;
    using System.Fabric.Common;
    using System.Fabric.Common.Tracing;
    using System.Globalization;
    using System.Linq;
    using System.Threading.Tasks;

    internal class TransactionManager
    {
        private readonly BeginTransactionOperationRatePerformanceCounterWriter beginTransactionOperationRatePerformanceCounterWriter;
        private readonly AddOperationRatePerformanceCounterWriter addOperationRatePerformanceCounterWriter;
        private readonly AddAtomicOperationRatePerformanceCounterWriter addAtomicOperationRatePerformanceCounterWriter;
        private readonly TransactionCommitRatePerformanceCounterWriter transactionCommitRatePerformanceCounterWriter;
        private readonly TransactionAbortRatePerformanceCounterWriter transactionAbortRatePerformanceCounterWriter;

        private readonly CheckpointManager checkpointManager;
        private readonly RecoveredOrCopiedCheckpointLsn recoveredOrCopiedCheckpointLsn;
        private readonly ReplicatedLogManager replicatedLogManager;
        private readonly TransactionMap transactionMap;
        private readonly OperationProcessor unlockCallbackManager;
        private readonly long flushAtBufferedBytes;
        private readonly ITracer tracer;

        public TransactionManager(
            RecoveredOrCopiedCheckpointLsn recoveredOrCopiedCheckpointLsn,
            TransactionMap transactionMap,
            FabricPerformanceCounterSetInstance instance,
            CheckpointManager checkpointManager,
            ReplicatedLogManager replicatedLogManager,
            OperationProcessor unlockCallbackManager,
            ITracer tracer,
            long flushAtBufferedBytes)
        {
            this.recoveredOrCopiedCheckpointLsn = recoveredOrCopiedCheckpointLsn;
            this.transactionMap = transactionMap;
            this.checkpointManager = checkpointManager;
            this.replicatedLogManager = replicatedLogManager;
            this.unlockCallbackManager = unlockCallbackManager;
            this.tracer = tracer;
            this.flushAtBufferedBytes = flushAtBufferedBytes;

            this.beginTransactionOperationRatePerformanceCounterWriter =
                new BeginTransactionOperationRatePerformanceCounterWriter(instance);
            this.addOperationRatePerformanceCounterWriter =
                new AddOperationRatePerformanceCounterWriter(instance);
            this.transactionCommitRatePerformanceCounterWriter =
                new TransactionCommitRatePerformanceCounterWriter(instance);
            this.transactionAbortRatePerformanceCounterWriter =
                new TransactionAbortRatePerformanceCounterWriter(instance);
            this.addAtomicOperationRatePerformanceCounterWriter =
                new AddAtomicOperationRatePerformanceCounterWriter(instance);
        }

        public TransactionMap TransactionsMap
        {
            get { return this.transactionMap; }
        }

        public void CreateTransactionState(
            Epoch logTailEpoch,
            TransactionLogRecord record)
        {
            switch (record.RecordType)
            {
                case LogRecordType.BeginTransaction:
                    var beginTxRecord = record as BeginTransactionOperationLogRecord;

                    this.beginTransactionOperationRatePerformanceCounterWriter.Increment();

                    if (beginTxRecord.IsSingleOperationTransaction)
                    {
                        this.transactionCommitRatePerformanceCounterWriter.Increment();

                        FabricEvents.Events.AcceptSingleOperationTransaction(
                            tracer.Type,
                            beginTxRecord.Transaction.Id,
                            record.Lsn.LSN);

                        record.Transaction.CommitSequenceNumber = record.Lsn.LSN;
                    }
                    else
                    {
                        FabricEvents.Events.AcceptBeginTransaction(
                            tracer.Type,
                            beginTxRecord.Transaction.Id,
                            record.Lsn.LSN);
                    }

                    beginTxRecord.RecordEpoch = logTailEpoch;
                    this.transactionMap.CreateTransaction(beginTxRecord);
                    break;

                case LogRecordType.Operation:
                    var dataRecord = record as OperationLogRecord;

                    if (dataRecord.Transaction.IsAtomicOperation || dataRecord.IsRedoOnly)
                    {
                        this.addAtomicOperationRatePerformanceCounterWriter.Increment();
                        dataRecord.Transaction.CommitSequenceNumber = record.Lsn.LSN;
                    }
                    else
                    {
                        this.addOperationRatePerformanceCounterWriter.Increment();
                    }

                    FabricEvents.Events.AcceptOperation(tracer.Type, dataRecord.Transaction.Id, record.Lsn.LSN);

                    this.transactionMap.AddOperation(dataRecord);
                    Utility.Assert(
                        (dataRecord.IsEnlistedTransaction == true)
                        || (dataRecord.Lsn <= this.recoveredOrCopiedCheckpointLsn.Value),
                        "(record.IsEnlistedTransaction == true) || (record.LastLogicalSequenceNumber <= this.recoveredOrCopiedCheckpointLsn). record LSN :{0} and recoverd or copied checkpoint lsn:{1}. record.InEnlistedTx :{2}",
                        dataRecord.Lsn, recoveredOrCopiedCheckpointLsn,
                        dataRecord.IsEnlistedTransaction);
                    break;

                case LogRecordType.EndTransaction:
                    var endTxRecord = record as EndTransactionLogRecord;

                    if (endTxRecord.IsCommitted)
                    {
                        this.transactionCommitRatePerformanceCounterWriter.Increment();
                        endTxRecord.Transaction.CommitSequenceNumber = record.Lsn.LSN;
                    }
                    else
                    {
                        this.transactionAbortRatePerformanceCounterWriter.Increment();
                    }

                    FabricEvents.Events.AcceptEndTransaction(
                        tracer.Type,
                        endTxRecord.Transaction.Id,
                        record.Lsn.LSN,
                        endTxRecord.IsCommitted);

                    this.transactionMap.CompleteTransaction(endTxRecord);
                    Utility.Assert(
                        (endTxRecord.IsEnlistedTransaction == true)
                        || (record.Lsn <= this.recoveredOrCopiedCheckpointLsn.Value),
                        "(record.IsEnlistedTransaction == true) || (record.LastLogicalSequenceNumber <= this.recoveredOrCopiedCheckpointLsn). record LSN :{0} and recoverd or copied checkpoint lsn:{1}. record.InEnlistedTx :{2}",
                        record.Lsn,
                        this.recoveredOrCopiedCheckpointLsn.Value.LSN,
                        endTxRecord.IsEnlistedTransaction);
                    break;

                default:
                    Utility.CodingError("Unexpected record type in TransactionsManager CreateTransactionState {0}", record.RecordType);
                    break;
            }
        }

        public async Task<long> AbortTransactionAsync(Transaction transaction)
        {
            LogRecord endTransactionRecord = await this.EndTransactionAsync(transaction, false, true).ConfigureAwait(false);
            return endTransactionRecord.Lsn.LSN;
        }

        public void AddOperation(
            Transaction transaction,
            OperationData metaData,
            OperationData undo,
            OperationData redo,
            object operationContext)
        {
            var record = new OperationLogRecord(transaction, metaData, undo, redo, operationContext);
            this.ProcessLogicalRecordOnPrimary(record);
        }

        public async Task<long> AddOperationAsync(
            AtomicOperation atomicOperation,
            OperationData metaData,
            OperationData undo,
            OperationData redo,
            object operationContext)
        {
            var record = new OperationLogRecord(atomicOperation, metaData, undo, redo, operationContext);
            await this.ProcessLogicalRecordOnPrimaryAsync(record).ConfigureAwait(false);

            return record.Lsn.LSN;
        }

        public async Task<long> AddOperationAsync(
            AtomicRedoOperation atomicRedoOperation,
            OperationData metaData,
            OperationData redo,
            object operationContext)
        {
            var record = new OperationLogRecord(atomicRedoOperation, metaData, redo, operationContext);
            await this.ProcessLogicalRecordOnPrimaryAsync(record).ConfigureAwait(false);

            return record.Lsn.LSN;
        }

        public void BeginTransaction(Transaction transaction, OperationData metaData, OperationData undo, OperationData redo, object operationContext)
        {
            var record = new BeginTransactionOperationLogRecord(
                transaction,
                metaData,
                undo,
                redo,
                operationContext,
                isSingleOperationTransaction: false);
            this.ProcessLogicalRecordOnPrimary(record);
        }

        public async Task<long> BeginTransactionAsync(Transaction transaction, OperationData metaData, OperationData undo, OperationData redo, object operationContext)
        {
            var record = new BeginTransactionOperationLogRecord(
                transaction,
                metaData,
                undo,
                redo,
                operationContext,
                isSingleOperationTransaction: true);
            await this.ProcessLogicalRecordOnPrimaryAsync(record).ConfigureAwait(false);
            return record.Lsn.LSN;
        }

        public async Task<long> CommitTransactionAsync(Transaction transaction)
        {
            LogRecord endTransactionRecord = await this.EndTransactionAsync(transaction, true, true).ConfigureAwait(false);
            return endTransactionRecord.Lsn.LSN;
        }

        private void InitiateLogicalRecordProcessingOnPrimary(LogicalLogRecord record)
        {
            this.checkpointManager.ThrowIfThrottled(record);

            long bufferedRecordsSizeBytes = 0;

            switch (record.RecordType)
            {
                case LogRecordType.BeginTransaction:
                case LogRecordType.Operation:
                case LogRecordType.EndTransaction:
                    bufferedRecordsSizeBytes = this.replicatedLogManager.ReplicateAndLog(record, this);
                    break;

                default:
                    Utility.CodingError("Unexpected record type {0} in InitiateLogicalRecordProcessingOnPrimary", record.RecordType);
                    break;
            }

            var addedIndexingRecord = false;
            var addedTruncateHeadRecord = false;

            this.checkpointManager.InsertPhysicalRecordsIfNecessary(out addedIndexingRecord, out addedTruncateHeadRecord);

            if (record.IsLatencySensitiveRecord == true || addedTruncateHeadRecord)
            {
                this.checkpointManager.RequestGroupCommit();
            }
            else if (bufferedRecordsSizeBytes > this.flushAtBufferedBytes)
            {
                {
                    // If there is lot of buffered data, issue a flush
                    var t =
                        this.replicatedLogManager.LogManager.FlushAsync(
                            "InitiateLogicalLogRecordProcessing BufferedRecordsSize: " + bufferedRecordsSizeBytes);
                }
            }

            return;
        }

        internal async Task<EndTransactionLogRecord> EndTransactionAsync(
            TransactionBase transaction,
            bool isCommitted,
            bool isThisReplicaTransaction)
        {
            var record = new EndTransactionLogRecord(
                transaction,
                isCommitted,
                isThisReplicaTransaction);
            await this.ProcessLogicalRecordOnPrimaryAsync(record).ConfigureAwait(false);

            return record;
        }

        private void ProcessLogicalRecordOnPrimary(TransactionLogRecord record)
        {
            this.ReplicateAndLogLogicalRecord(record);

            Task.WhenAll(record.AwaitApply(), LoggingReplicator.AwaitReplication(record, this.replicatedLogManager)).IgnoreException().ContinueWith(
                task =>
                {
                    // Simply retrieve task exception to mark it as handled
                    if (task.Exception != null)
                    {
                        Utility.Assert(task.IsFaulted == true, "task.IsFaulted == true");
                    }

                    this.InvokeUnlockCallback(record);
                }).IgnoreExceptionVoid();
        }

        /// <summary>
        /// Primary: Process the logical log record.
        /// </summary>
        /// <param name="record"></param>
        /// <returns></returns>
        /// <remarks>
        /// Wait for the logical record to be flushed and replicated.
        /// If it is a barrier, increase stable lsn.
        /// If it is a operation record, unlock it.
        /// </remarks>
        private async Task ProcessLogicalRecordOnPrimaryAsync(TransactionLogRecord record)
        {
            Exception exception = null;
            this.ReplicateAndLogLogicalRecord(record);

            try
            {
                // Wait for the record to be flushed and replicated
                await Task.WhenAll(record.AwaitApply(), LoggingReplicator.AwaitReplication(record, this.replicatedLogManager)).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                exception = e;
            }
            finally
            {
                this.InvokeUnlockCallback(record);
            }

            // Throw TransactionFaultedException if there has been any failure during replication and apply awaits.
            if (exception != null)
            {
                throw new TransactionFaultedException(exception.Message);
            }

            return;
        }

        private void ReplicateAndLogLogicalRecord(TransactionLogRecord record)
        {
            this.InitiateLogicalRecordProcessingOnPrimary(record);
            return;
        }

        private void InvokeUnlockCallback(TransactionLogRecord record)
        {
            this.unlockCallbackManager.Unlock(record);
        }
    }
}

File: src/prod/src/data/txnreplicator/loggingreplicator/TransactionManager.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LoggingReplicator;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

Common::StringLiteral const TraceComponent("TransactionManager");
Common::TimeSpan const TransactionManager::ConfigUpdateIntervalInSeconds = Common::TimeSpan::FromSeconds(30);

TransactionManager::TransactionManager(
    __in PartitionedReplicaId const & traceId,
    __in RecoveredOrCopiedCheckpointState const & recoveredOrCopiedCheckpointLsn,
    __in TransactionMap & transactionMap,
    __in CheckpointManager & checkpointManager,
    __in ReplicatedLogManager & replicatedLogManager,
    __in OperationProcessor & operationProcessor,
    __in InvalidLogRecords & invalidLogRecords,
    __in TRInternalSettingsSPtr const & transactionalReplicatorConfig,
    __in TRPerformanceCountersSPtr const & perfCounters)
    : KObject()
    , KShared()
    , PartitionedReplicaTraceComponent(traceId)
    , transactionalReplicatorConfig_(transactionalReplicatorConfig)
    , configUpdateStopWatch_(Common::Stopwatch())
    , recoveredOrCopiedCheckpointState_(&recoveredOrCopiedCheckpointLsn)
    , transactionMap_(&transactionMap)
    , checkpointManager_(&checkpointManager)
    , replicatedLogManager_(&replicatedLogManager)
    , operationProcessor_(&operationProcessor)
    , invalidLogRecords_(&invalidLogRecords)
    , perfCounters_(perfCounters)
{
    EventSource::Events->Ctor(
        TracePartitionId,
        ReplicaId,
        L"TransactionManager",
        reinterpret_cast<uintptr_t>(this));

    configUpdateStopWatch_.Start();
}

TransactionManager::~TransactionManager()
{
    EventSource::Events->Dtor(
        TracePartitionId,
        ReplicaId,
        L"TransactionManager",
        reinterpret_cast<uintptr_t>(this));
}

TransactionManager::SPtr TransactionManager::Create(
    __in PartitionedReplicaId const & traceId,
    __in RecoveredOrCopiedCheckpointState const & recoveredOrCopiedCheckpointLsn,
    __in TransactionMap & transactionMap,
    __in CheckpointManager & checkpointManager,
    __in ReplicatedLogManager & replicatedLogManager,
    __in OperationProcessor & operationProcessor,
    __in InvalidLogRecords & invalidLogRecords,
    __in TRInternalSettingsSPtr const & transactionalReplicatorConfig,
    __in TRPerformanceCountersSPtr const & perfCounters,
    __in KAllocator & allocator)
{
    TransactionManager * pointer = _new(TRANSACTIONMANAGER_TAG, allocator) TransactionManager(
        traceId,
        recoveredOrCopiedCheckpointLsn,
        transactionMap,
        checkpointManager,
        replicatedLogManager,
        operationProcessor,
        invalidLogRecords,
        transactionalReplicatorConfig,
        perfCounters);

    THROW_ON_ALLOCATION_FAILURE(pointer);

    return TransactionManager::SPtr(pointer);
}

void TransactionManager::CreateTransactionState(
    __in Epoch const & tailEpoch,
    __in TransactionLogRecord & record) noexcept
{
    TransactionLogRecord::SPtr recordSPtr = &record;

    switch(recordSPtr->RecordType)
    {
    case LogRecordType::Enum::BeginTransaction:
    {
        BeginTransactionOperationLogRecord::SPtr beginTxRecord = dynamic_cast<BeginTransactionOperationLogRecord *>(recordSPtr.RawPtr());
        ASSERT_IF(
            beginTxRecord == nullptr,
            "{0}: Unexpected dynamic cast failure",
            TraceId);

        perfCounters_->BeginTransactionOperationRate.Increment();

        if (beginTxRecord->IsSingleOperationTransaction)
        {
            perfCounters_->CommitTransactionRate.Increment();

            EventSource::Events->AcceptSingleOperationTransaction(
                TracePartitionId,
                ReplicaId,
                beginTxRecord->BaseTransaction.TransactionId,
                recordSPtr->Lsn);
        }
        else
        {
            EventSource::Events->AcceptBeginTransaction(
                TracePartitionId,
                ReplicaId,
                beginTxRecord->BaseTransaction.TransactionId,
                recordSPtr->Lsn);
        }

        beginTxRecord->RecordEpoch = tailEpoch;
        transactionMap_->CreateTransaction(*beginTxRecord);

        break;
    }

    case LogRecordType::Enum::Operation: 
    {
        OperationLogRecord::SPtr dataRecord = dynamic_cast<OperationLogRecord *>(recordSPtr.RawPtr());
        ASSERT_IF(
            dataRecord == nullptr,
            "{0}: Unexpected dynamic cast failure",
            TraceId);

        if (dataRecord->BaseTransaction.IsAtomicOperation || dataRecord->IsRedoOnly)
        {
            perfCounters_->AtomicOperationRate.Increment();

            EventSource::Events->AcceptAtomicOperation(
                TracePartitionId,
                ReplicaId,
                dataRecord->BaseTransaction.TransactionId,
                recordSPtr->Lsn,
                dataRecord->IsRedoOnly);
        }
        else
        {
            perfCounters_->AddOperationRate.Increment();

            EventSource::Events->AcceptOperation(
                TracePartitionId,
                ReplicaId,
                dataRecord->BaseTransaction.TransactionId,
                recordSPtr->Lsn);
        }

        transactionMap_->AddOperation(*dataRecord);

        ASSERT_IFNOT(
            dataRecord->IsEnlistedTransaction ||
            dataRecord->Lsn <= recoveredOrCopiedCheckpointState_->Lsn,
            "{0}: Invalid op xact log record - lsn: {1}, recoverd or copied checkpoint lsn: {2}, enlisted: {3}",
            TraceId,
            dataRecord->Lsn,
            recoveredOrCopiedCheckpointState_->Lsn,
            dataRecord->IsEnlistedTransaction);

        break;
    }
    case LogRecordType::Enum::EndTransaction:
    {
        EndTransactionLogRecord::SPtr endTxRecord = dynamic_cast<EndTransactionLogRecord *>(recordSPtr.RawPtr());
        ASSERT_IF(
            endTxRecord == nullptr,
            "{0}: Unexpected dynamic cast failure",
            TraceId);

        if (endTxRecord->IsCommitted)
        {
            perfCounters_->CommitTransactionRate.Increment();
        }
        else 
        {
            perfCounters_->AbortTransactionRate.Increment();
        }

        EventSource::Events->AcceptEndTransaction(
            TracePartitionId,
            ReplicaId,
            endTxRecord->BaseTransaction.TransactionId,
            recordSPtr->Lsn,
            endTxRecord->IsCommitted);

        transactionMap_->CompleteTransaction(*endTxRecord);

        ASSERT_IFNOT(
            endTxRecord->IsEnlistedTransaction ||
            endTxRecord->Lsn <= recoveredOrCopiedCheckpointState_->Lsn,
            "{0}: Invalid end xact log record - lsn: {1}, recoverd or copied checkpoint lsn: {2}, enlisted: {3}",
            TraceId,
            endTxRecord->Lsn,
            recoveredOrCopiedCheckpointState_->Lsn,
            endTxRecord->IsEnlistedTransaction);

        break;
    }
    default: 
        
        ASSERT_IFNOT(
            false,
            "{0}: CreateTransactionState | Invalid record type {0}",
            TraceId,
            record.RecordType);
        break;
    }
}

Awaitable<NTSTATUS> TransactionManager::AbortTransactionAsync(
    __in Transaction & transaction,
    __out LONG64 & abortLsn) noexcept
{
    abortLsn = Constants::InvalidLsn;

    EndTransactionLogRecord::SPtr result;
    NTSTATUS status = co_await EndTransactionAsync(transaction, false, true, result);

    CO_RETURN_ON_FAILURE(status);

    abortLsn = result->Lsn;

    co_return status;
}

NTSTATUS TransactionManager::AddOperation(
    __in Transaction & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext) noexcept
{
    OperationLogRecord::SPtr record = OperationLogRecord::Create(
        transaction,
        metaData,
        undo,
        redo,
        operationContext,
        *invalidLogRecords_->Inv_PhysicalLogRecord,
        *invalidLogRecords_->Inv_TransactionLogRecord,
        GetThisAllocator());
    
    NTSTATUS status = ReplicateAndLogLogicalRecord(*record);

    RETURN_ON_FAILURE(status);

    ProcessLogicalRecordOnPrimary(*record);

    return STATUS_SUCCESS;
}

Awaitable<NTSTATUS> TransactionManager::AddOperationAsync(
    __in AtomicOperation & atomicOperation,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __out LONG64 & commitLsn) noexcept
{
    commitLsn = Constants::InvalidLsn;

    OperationLogRecord::SPtr record = OperationLogRecord::Create(
        atomicOperation,
        metaData,
        undo,
        redo,
        operationContext,
        *invalidLogRecords_->Inv_PhysicalLogRecord,
        *invalidLogRecords_->Inv_TransactionLogRecord,
        GetThisAllocator());

    NTSTATUS status = ReplicateAndLogLogicalRecord(*record);

    CO_RETURN_ON_FAILURE(status);

    status = co_await ProcessLogicalRecordOnPrimaryAsync(*record);

    if (NT_SUCCESS(status))
    {
        commitLsn = record->Lsn;
    }

    co_return status;
}

Awaitable<NTSTATUS> TransactionManager::AddOperationAsync(
    __in AtomicRedoOperation & atomicRedoOperation,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __out LONG64 & commitLsn) noexcept
{
    commitLsn = Constants::InvalidLsn;

    OperationLogRecord::SPtr record = OperationLogRecord::Create(
        atomicRedoOperation,
        metaData,
        redo,
        operationContext,
        *invalidLogRecords_->Inv_PhysicalLogRecord,
        *invalidLogRecords_->Inv_TransactionLogRecord,
        GetThisAllocator());

    NTSTATUS status = ReplicateAndLogLogicalRecord(*record);
    
    CO_RETURN_ON_FAILURE(status);

    status = co_await ProcessLogicalRecordOnPrimaryAsync(*record);

    if (NT_SUCCESS(status))
    {
        commitLsn = record->Lsn;
    }

    co_return status;
}

NTSTATUS TransactionManager::BeginTransaction(
    __in Transaction & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext) noexcept
{
    BeginTransactionOperationLogRecord::SPtr record = BeginTransactionOperationLogRecord::Create(
        transaction,
        false, // isSingleOperationTransaction
        metaData,
        undo,
        redo,
        operationContext,
        *invalidLogRecords_->Inv_PhysicalLogRecord,
        *invalidLogRecords_->Inv_TransactionLogRecord,
        GetThisAllocator());

    NTSTATUS status = ReplicateAndLogLogicalRecord(*record);

    RETURN_ON_FAILURE(status);

    ProcessLogicalRecordOnPrimary(*record);

    return STATUS_SUCCESS;
}

Awaitable<NTSTATUS> TransactionManager::BeginTransactionAsync(
    __in Transaction & transaction,
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in_opt OperationContext const * const operationContext,
    __out LONG64 & commitLsn) noexcept
{
    commitLsn = Constants::InvalidLsn;

    BeginTransactionOperationLogRecord::SPtr record = BeginTransactionOperationLogRecord::Create(
        transaction,
        true, // isSingleOperationTransaction
        metaData,
        undo,
        redo,
        operationContext,
        *invalidLogRecords_->Inv_PhysicalLogRecord,
        *invalidLogRecords_->Inv_TransactionLogRecord,
        GetThisAllocator());

    NTSTATUS status = ReplicateAndLogLogicalRecord(*record);

    CO_RETURN_ON_FAILURE(status);

    status = co_await ProcessLogicalRecordOnPrimaryAsync(*record);

    if (NT_SUCCESS(status))
    {
        commitLsn = record->Lsn;
    }

    co_return status;
}

Awaitable<NTSTATUS> TransactionManager::CommitTransactionAsync(
    __in Transaction & transaction,
    __out LONG64 & commitLsn) noexcept
{
    commitLsn = Constants::InvalidLsn;

    EndTransactionLogRecord::SPtr record = nullptr;

    NTSTATUS status = co_await EndTransactionAsync(
        transaction, 
        true, 
        true, 
        record);
    
    CO_RETURN_ON_FAILURE(status);

    commitLsn = record->Lsn;
    co_return status;
}

NTSTATUS TransactionManager::InitiateLogicalRecordProcessingOnPrimary(__in LogicalLogRecord & record) noexcept
{
    NTSTATUS status = checkpointManager_->ErrorIfThrottled(record);

    RETURN_ON_FAILURE(status);

    LogicalLogRecord::SPtr recordSPtr = &record;
    LONG64 bufferedRecordsSizebytes = 0;

    switch (recordSPtr->RecordType)
    {
    case LogRecordType::Enum::BeginTransaction:
    case LogRecordType::Enum::Operation:
    case LogRecordType::Enum::EndTransaction:
        status = replicatedLogManager_->ReplicateAndLog(*recordSPtr, bufferedRecordsSizebytes, this);
        break;
    default:
        ASSERT_IFNOT(
            false,
            "{0}: InitiateLogicalRecordProcessingOnPrimary | Invalid record type {1}",
            TraceId,
            record.RecordType);
        break;
    }

    RETURN_ON_FAILURE(status);

    bool addedIndexingRecord = false;
    bool addedTruncateHeadRecord = false;

    checkpointManager_->InsertPhysicalRecordsIfNecessary(
        addedIndexingRecord,
        addedTruncateHeadRecord);

    if (recordSPtr->IsLatencySensitiveRecord || addedTruncateHeadRecord)
    {
        checkpointManager_->RequestGroupCommit();
    }
    else if (bufferedRecordsSizebytes > (transactionalReplicatorConfig_->MaxRecordSizeInKB * 1024))
    {
        Awaitable<void> flushTask = replicatedLogManager_->InnerLogManager->FlushAsync(L"InitiateLogicalRecordProcessingOnPrimary");
        ToTask(flushTask);
    }

    return status;
}

Awaitable<NTSTATUS> TransactionManager::EndTransactionAsync(
    __in TransactionBase & transaction,
    __in bool isCommitted,
    __in bool isThisReplicaTransaction,
    __out EndTransactionLogRecord::SPtr & result) noexcept
{
    result = EndTransactionLogRecord::Create(
        transaction,
        isCommitted,
        isThisReplicaTransaction,
        *invalidLogRecords_->Inv_PhysicalLogRecord,
        *invalidLogRecords_->Inv_TransactionLogRecord,
        GetThisAllocator());

    NTSTATUS status = ReplicateAndLogLogicalRecord(*result);

    CO_RETURN_ON_FAILURE(status);

    status = co_await ProcessLogicalRecordOnPrimaryAsync(*result);

    co_return status;
}

// NOTE - In managed code, this method had the replicate call while in native we do not have it here
// This is because we cannot throw out of this function if beginreplicate fails as Task's dont expect an exception to be thrown
Task TransactionManager::ProcessLogicalRecordOnPrimary(__in TransactionLogRecord & record)
{
    KCoShared$ApiEntry();
    TransactionLogRecord::SPtr recordSPtr = &record;

    NTSTATUS status1 = co_await recordSPtr->AwaitApply();
    NTSTATUS status2 = co_await replicatedLogManager_->AwaitReplication(*recordSPtr);

    // Keeping these variables in case we need for debugging
    UNREFERENCED_PARAMETER(status1);
    UNREFERENCED_PARAMETER(status2);

    InvokeUnlockCallback(*recordSPtr);
    co_return;
}

Awaitable<NTSTATUS> TransactionManager::ProcessLogicalRecordOnPrimaryAsync(__in TransactionLogRecord & record) noexcept
{
    TransactionLogRecord::SPtr recordSPtr = &record;

    NTSTATUS status1 = co_await recordSPtr->AwaitApply();
    NTSTATUS status2 = co_await replicatedLogManager_->AwaitReplication(*recordSPtr);

    NTSTATUS status = FIRST_NTSTATUS_ERROR(status1, status2);

    // Unlock to ensure any pending operations are released
    InvokeUnlockCallback(*recordSPtr);

    if (!NT_SUCCESS(status))
    {
        // Use a SF_STATUS_TRANSACTION_ABORTED error code to maintain the exception contract where user can expect a specific set of errors
        co_return SF_STATUS_TRANSACTION_ABORTED;
    }

    co_return STATUS_SUCCESS;
}

NTSTATUS TransactionManager::ReplicateAndLogLogicalRecord(__in TransactionLogRecord & record) noexcept
{
    return InitiateLogicalRecordProcessingOnPrimary(record);
}

void TransactionManager::InvokeUnlockCallback(__in TransactionLogRecord & record) noexcept
{
    operationProcessor_->Unlock(record);
}
